biigle/core
# Vue 3

Tasks:

- [x] Install Vue 3 compat
- [x] Set up Vite build
- [x] Update uiv
- [x] Test Vite hot reload
- [x] Fix navbar dropdowns
- [x] Find Vue Compat warnings and fix them
   - Use fake privacy page for default layout/code
   - Disable compat features to find source of warnings
   - Walk through all views and look for warnings
- [x] Update slots to `<template #slot>` syntax
- [x] Update use of `v-model` to new default `modelValue` variable name
- [x] Define component events with `emits: ['evt', ...]`
- [x] Make image annotation tool work
- [x] Make video annotation tool work
   - [x] Check annotation model if any properties must be reactive
   - [x] Implement annotation session filter. Resolves https://github.com/biigle/core/issues/281
- [x] Replace use of `$on`, `$once` and `$off`.
- [x] Replace use of `inline-template`.
- [x] Update `:disabled` and `v-bind:disabled` attributes that expect the attribute to be removed if `false`. Use `null` instead.
- [x] Rename `beforeDestroy` to `beforeUnmount`.
- [x] Fix these locations:
   - [x] Error `Property "scollable" was accessed during render but is not defined on instance.` on project overview.
   - [x] The image/video annotation tab filter system. It uses `Vue.extend`, `$mount()`, `$on()` and `$off()`. Fix this once the image/video annotation tools are functional again with the compatibility build. Finally disable `GLOBAL_EXTEND`.
- [x] Remove/replace Vue Resource. Afterwards, disable `GLOBAL_PROTOTYPE`. Maybe use a [fetch wrapper](https://github.com/uiv-lib/uiv/blob/dev/src/utils/http.utils.js).
   - We keep vue-resource for now, as it also can be used directly (not as Vue plugin).
- [x] Replace Vue.Promise with native Promise
- [x] Walk through all JS files and test their features in the UI. View routes:
    - [x] admin/announcements
    - [x] admin/announcements/new
    - [x] admin/export
       - uiv Tabs component produces warning
    - [x] admin/federated-search
    - [x] admin/import
    - [x] admin/import/{token}
    - [x] admin/label-trees
    - [x] admin/logs
    - [x] admin/logs/{file}
    - [x] admin/users
    - [x] admin/users/delete/{id}
    - [x] admin/users/edit/{id}
    - [x] admin/users/new
    - [x] admin/users/{id}
    - [x] image-annotations/{id}
    - [x] images/{id}
    - [x] images/{id}/annotations
    - [x] imprint
    - [x] label-trees
    - [x] label-trees/create
    - [x] label-trees/import
    - [x] label-trees/{id}
       - uiv Tabs component produces warning
    - [x] label-trees/{id}/members
    - [x] label-trees/{id}/merge
    - [x] label-trees/{id}/merge/{id2}
    - [x] label-trees/{id}/projects
    - [x] label-trees/{id}/versions/create
    - [x] label-trees/{id}/versions/{id2}
    - [x] manual/documentation/{module}/{article?}
    - [x] manual/tutorials/{module}/{article?}
    - [x] pending-volumes/{id}
    - [x] pending-volumes/{id}/annotation-labels
    - [x] pending-volumes/{id}/file-labels
    - [x] pending-volumes/{id}/finish
    - [x] pending-volumes/{id}/label-map
    - [x] pending-volumes/{id}/user-map
    - [x] privacy
    - [x] project-invitations/{uuid}
    - [x] projects/create
    - [x] projects/{id}
    - [x] projects/{id}/charts
    - [x] projects/{id}/label-trees
    - [x] projects/{id}/members
    - [x] projects/{id}/reports
    - [x] settings/account
    - [x] settings/authentication
    - [x] settings/notifications
    - [x] settings/profile
    - [x] settings/tokens
    - [x] terms
    - [x] video-annotations/{id}
    - [x] videos/{id}
    - [x] videos/{id}/annotations
    - [x] volumes/clone/{id}
    - [x] volumes/create
    - [x] volumes/edit/{id}
    - [x] volumes/{id}
    - [x] volumes/{id}/reports
- [x] Update modules:
   - [x] biigle/largo https://github.com/biigle/largo/pull/206
   - [x] ~~biigle/metadata-ifdo~~
   - [x] biigle/user-storage https://github.com/biigle/user-storage/pull/55
   - [x] biigle/user-disks https://github.com/biigle/user-disks/pull/23
   - [x] ~~biigle/newsletter~~ https://github.com/biigle/newsletter/pull/18
   - [x] biigle/module https://github.com/biigle/module/pull/25
   - [x] ~~biigle/metadata-coco~~
   - [x] biigle/maia https://github.com/biigle/maia/pull/182
   - [x] biigle/magic-sam https://github.com/biigle/magic-sam/pull/49
   - [x] biigle/laserpoints https://github.com/biigle/laserpoints/pull/92
   - [x] ~~biigle/kpis~~
   - [x] biigle/geo https://github.com/biigle/geo/pull/65
   - [x] ~~biigle/denbi-service~~
   - [x] ~~biigle/demo~~
   - [x] biigle/color-sort https://github.com/biigle/color-sort/pull/56
   - [x] ~~biigle/auth-nfdi~~
   - [x] ~~biigle/auth-lslogin~~
   - [x] biigle/ananas https://github.com/biigle/ananas/pull/44
   - [x] Checklist
       - [Migrate from Mix to Vite](https://github.com/laravel/vite-plugin/blob/main/UPGRADE.md#migrating-from-laravel-mix-to-vite).
       - Fix errors returned by `npm run build`
       - Change `<script type="text/javascript">` to `<script type="module">` for inline scripts.
       - Update use of Vue Resource.
       - Update `<control-button>` component to use scoped slot event instead of `$on` listener ([689106d50](https://github.com/biigle/core/commit/689106d504342b5c20e009c5282bfccada6aa425)).
       - Update slots `<template slot="xxx">` to `<template #xxx>` syntax.
       - Update use of `v-model` to new default `modelValue` variable name.
       - Define component events with `emits: ['evt', ...]`.
       - Replace use of `$on`, `$once` and `$off`.
       - Replace use of `inline-template`.
       - Update `:disabled` and `v-bind:disabled` attributes that expect the attribute to be removed if `false`. Use `null` instead.
       - Rename `beforeDestroy` to `beforeUnmount`.
       - Replace Vue.Promise with native Promise.
       - Fix all warnings of Vue Compat.
       - Fix import of `handleErrorResponse`.
       - Test all features.
- [x] Update Docker build process (if necessary) https://github.com/biigle/biigle/pull/34
- [x] ~~Maybe remove `cachebust_asset()`~~ (still useful)
- [x] ~~Consider to share OpenLayers similar to Vue between core and modules (e.g. magic-sam, reports, ananas, geo, ...)~~
    - This can't be done the way OpenLayers is built. First, there are no "dev" and "production" builds of OpenLayers. These could be created manually but: OpenLayers components are imported via specifiers such as `ol/Map`. With an importmap, all specifies would need to be defined, not only one as with Vue. This is just not practical, so OL will be imported individually in modules and critical components (such as styles) will be shared with the import/export system of BIIGLE.
- [x] TEST EVERYTHING with compat `MODE: 3`
  - [x] Change `<script type="text/javascript">` to `<script type="module">` for inline scripts that were added in the meantime
  - [x] Check all view routes again.
  - [x] Check if date picker components behave correctly
  - [x] Check if echarts charts work correctly
- [x] Invite everyone to test the UI and find errors
- [x] Remove Vue Compat (`configureCompat` and `TODO`s), including all uses of `compatConfig` (also in modules)
- [x] Test Vue2 vs Vue3 performance on huge video/image with lots of annotations
- [ ] Merge the current master/main of core and modules and check the updated code for compatibility.
- [ ] Delete dev-modules-vue3 branch.
- [ ] Check if #984 is resolved
- [ ] Follow up with #249 (unify image and video annotation tools with composition API)

Resolves #455 
Resolves #515
